### ISSUE TITLE
Installation failed - mamba not found

### ISSUE THREAD

**Issue by eyalbenda** [bcbio/bcbio-nextgen#3234]
My installation fails on two separate clusters, due to the mamba binary not found (trying the path in the error indeed doesn't find mamba).
I'm attaching output and error logs.

 

**Version info**
 - bcbio version (bcbio_nextgen.py --version):
 - OS name and version: CentOS 6.6 and 6.10

**To Reproduce**
 - Exact bcbio command you have used: python bcbio_nextgen_install.py.1 --cores 20 --nodata --isolate --distribution centos -u development ~/bcbio_nextgen --tooldir ~/bcbio_nextgen/tools

**Observed behavior**
[log_output.txt](https://github.com/bcbio/bcbio-nextgen/files/4672863/log_output.txt)
[log_error.txt](https://github.com/bcbio/bcbio-nextgen/files/4672864/log_error.txt)

**Comment by eyalbenda** [comment]
I found out the problem. This is due to me trying to install bcbio_nextgen while having an active Conda environment. This didn't use to be a problem, but it is now. I'm still getting stuck since I'm not able to use mamba on this old cluster (mamba install just has no output). Is there a way to revert to using Conda?

**Comment by eyalbenda** [comment]
I changed the obvious parts of the python code to just use conda (setting mamba_bin to "conda" always). I suggest this becomes a flag. 
I'm now able to progress in the install. Please let me know if you'd like me to close the issue, or if you'd consider addressing these issues. summary:

- It is impossible to install bcbio if a conda environment is activated. In my case, I needed to build python from source since the cluster had python 2.6 installed.
- mamba, which doesn't function on this old cluster, breaks the installation.

**Comment by naumenko-sa** [comment]
Hi Eyal @eyalbenda !

Thanks for reporting!

We are using mamba to speed up bcbio installation. The installer creates several environments and with conda it takes up to 24h to solve and install. With mamba it is 30 min. 

Bcbio does not rely on mamba or conda installed in the system - it installs its own mamba, conda, and python. But you need at least python2.7 with some minimal dependencies to kickstart the bcbio installer script. Maybe your cluster has a module system, `module load python/python_version` to load a more fresh python?

Mixing conda environments could create more issues in the future when you are running bcbio, it is better to avoid that. Unfortunately, we cannot create nested conda environments. So you need to deactivate your conda environment and also you may need to cleanup `~/.conda/environemnts.txt` before running the bcbio install. A good point - we need to clarify that in the docs.

Why mamba is not running on the old cluster?

Sergey

**Comment by eyalbenda** [comment]
Thanks Sergey!
I have no idea why mamba doesn't work, I opened an issue at that repository. For now, the installation is slowly running along. the version of git on the cluster was also too old so cloning the repo also didn't work, but I copied the repo into a local directory and changed the installer to reflect that as well. I hope there are no more surprises and that it works, even if it takes 24h.

**Comment by naumenko-sa** [comment]
thanks, I see it [here](https://github.com/QuantStack/mamba/issues/293).
Closing for now. Feel free to re-open if you still see installation issues!

**Comment by BrianLohman** [comment]
Hello @eyalbenda and @naumenko-sa,

Did deactivating conda and removing the bcbio env from your ```~/.conda/envrionments.txt``` work for you? I have encountered what is likely the same issue and those actions did not resolve it. I am on a CentOS 7 machine. 

I have tried:
```
mkdir bcbio # note that I have to make this dir manually as running the install with the /bcbio fails: cannot mkdir
python bcbio_nextgen_install.py ./bcbio --tooldir=./bcbio/tools --nodata
```

Full logs attached, but the error is:
```
An unexpected error has occurred. Conda has prepared the above report.

Traceback (most recent call last):
  File "bcbio_nextgen_install.py", line 290, in <module>
    main(parser.parse_args(), sys.argv[1:])
  File "bcbio_nextgen_install.py", line 44, in main
    subprocess.check_call([anaconda["mamba"], "install", "--yes", "conda-build"])
  File "/usr/lib64/python2.7/subprocess.py", line 542, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/bin/mamba', 'install', '--yes', 'conda-build']' returned non-zero exit status 1

```

Any help would be appreciated
[bcbio_install_logs.txt](https://github.com/bcbio/bcbio-nextgen/files/5680666/bcbio_install_logs.txt)

**Comment by roryk** [comment]
Hi Brian,

Sorry about that-- what does `/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/pkgs/cache/976d321b.json` look like? It looks like that is the error mamba is running into:

```
 >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/conda/exceptions.py", line 1079, in __call__
        return func(*args, **kwargs)
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/mamba/mamba.py", line 882, in exception_converter
        raise e
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/mamba/mamba.py", line 876, in exception_converter
        exit_code = _wrapped_main(*args, **kwargs)
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/mamba/mamba.py", line 835, in _wrapped_main
        result = do_call(args, p)
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/mamba/mamba.py", line 716, in do_call
        exit_code = install(args, parser, "install")
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/mamba/mamba.py", line 514, in install
        index = load_channels(pool, channels, repos)
      File "/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/lib/python3.7/site-packages/mamba/utils.py", line 135, in load_channels
        repo = subdir.create_repo(pool)
    RuntimeError: Could not read JSON repodata file (/uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/pkgs/cache/976d321b.json) parse error line 1
```

I think something is wrong with that file, what does it look like?

**Comment by roryk** [comment]
https://github.com/mamba-org/mamba/issues/512 has a similar issue and is due to running out of space, causing the repo file to get truncated. Do you know if that is happening to you? I don't see an out of space error in the log, but maybe you are running into quota issues?

**Comment by BrianLohman** [comment]
Hi @roryk,

Thanks very much for you quick reply. Here are the contents of the json file in questions:

```
cat /uufs/chpc.utah.edu/common/HIPAA/u0806040/src/bcbio/anaconda/pkgs/cache/976d321b.json
{"_cache_control":"public, max-age=1200","_etag":"W/\"14d906118b457546050faef47edac702-3\"","_mod":"Fri, 11 Dec 2020 15:24:26 GMT","_url":"https://conda.anaconda.org/bioconda/linux-64/repodata.json",[u0806040@redwood11 src]$
```
I double checked in vi and it looks like this file could be truncated? There is no right curly brace to complete the syntax. 

I also tried installing in the unlimited scratch directory and I got the same result: failure at the same place with the same message referencing this json in the cache that could be truncated

Thanks for your help

**Comment by BrianLohman** [comment]
Hi @roryk,

I think I found the cause: my bashrc sets my tmpdir on login as ```/scratch/local/[username]``` which did not exist yet (old bashrc carried over to a new machine). I created that directory and tried again. The installation is still in progress but is past where I was hung up.

Of course no space left on device and target directory DNE should produce similar errors. 

Thanks again for your help.
